src/topology/ready_arrays.rs

```diff
@@ -0,0 +1,86 @@
+use std::pin::Pin;
+
+use futures::{
+    task::{Context, Poll},
+    {Stream, StreamExt},
+};
+
+use crate::event::{EventArray, EventContainer};
+
+const DEFAULT_CAPACITY: usize = 4096;
+
+/// A stream combinator aimed at improving the performance of event streams under load.
+///
+/// This is similar in spirit to `StreamExt::ready_chunks`, but built specifically `EventArray`.
+/// The more general `FoldReady` is left as an exercise to the reader.
+pub struct ReadyArrays<T> {
+    inner: T,
+    enqueued: Vec<EventArray>,
```

Comment:
Did you consider having a single array that events are pushed into instead of an array of arrays?

Reply:
That should be straightforward to experiment with. It would also help with the concern around `DEFAULT_CAPACITY`.

Reply:
Might consider [this](https://github.com/vectordotdev/vector/pull/11695) as a starting point.

Reply:
I've clarified the documentation with regard to this question in [17b52be](https://github.com/vectordotdev/vector/pull/12626/commits/17b52bebab9576aa0f4d0006677d8ca673709e08).

Reply:
Aha! I knew there had been a `ready_arrays` somewhere but couldn't find it. And yes, it's likely simpler to accumulate into a single array. There is the danger of more copies and potential reallocation as the arrays grow, but that's something we can measure.

Reply:
Almost all of our limits when it comes to channel/buffer sizing are meant to be soft limits, so it's not clear why we'd want to pay the cost of reallocating when we don't _need_ to be so precise.

Where I could see this being useful is when event arrays are small and we end up wanting to batch them together to be more efficient about work done per unit... although I'd probably stump for a more intrusive collection-esque approach of chaining the smaller, individual arrays together and having an `EventContainer` implementation/adapter to make it work downstream.

Reply:
All of that said, for the purpose of how it's being used here, such an intrusive collection-esque design might have limited to no benefit, since the transform code can already operate on each individual batch, so there's no need to generate a single `Iterator<Item = Event>` from the individual arrays.

Reply:
Follow up: https://github.com/vectordotdev/vector/issues/12663